client/tracing/src/lib.rs

```diff
@@ -291,9 +290,6 @@ fn send_telemetry(span_datum: SpanDatum) {
 	);
 }
 
-fn send_grafana(span_datum: SpanDatum) {
-	let name = format!("{}::{}", span_datum.target, span_datum.name);
-	if let Err(e) = record_metrics!(&name => span_datum.overall_time.as_nanos(),) {
-		log::warn!("Unable to send metrics to grafana: {:?}", e);
-	}
+fn send_prometheus(span_datum: SpanDatum) {
+	unimplemented!()
 }
```

Comment:
@mxinden do you know of a way that we could implement this API? It seems like prometheus gauges etc are intended to created as static refs etc.

Reply:
My guess is that you can't implement it without creating a macro.

Or you can simply make a static reference.
https://github.com/tikv/rust-prometheus/blob/master/examples/example_edition_2018.rs

Reply:
I don't yet understand fully how prometheus works, but to fully support this I think we would need to be able to: 

- batch together all values (spans) collected since last scrape 
- partition by target and name (by the use of labels perhaps).  

With all runtime targets currently in the code enabled, we can be approaching 200 measurements per second just across the runtime.  Is this feasible to do (while preserving timestamp) - the alternative would be to consider to lose some detail and aggregate the values between scrapes, which would fix the first point above.

Reply:
Hmm. I'll have a go at this later.

Reply:
> be approaching 200 measurements

@mattrutherford 200 measurements are not an issue for Prometheus. Prometheus can easily handle couple of mega bytes of metric data per scrape. See https://github.com/kubernetes/kube-state-metrics/issues/498 for some numbers.

> batch together all values (spans) collected since last scrape

This sounds more like we want a Histogram, right?

---

Would it be possible to tackle the effort within `client/tracing` as a follow up pull request? I would like to keep this one small. What do you think @mattrutherford @expenses @nodebreaker0-0?

Reply:
a follow up PR sounds sensible -  am I right in thinking that Histogram does not preserve individual observations? If so, that's not really what we want, because eg. there can be a particular individual timing of a span that we care about, with a specific set of inputs that occurs at a particular time and we want to preserve all that. Eg we can create a span as such:

```rust
let span = tracing::span!(
	tracing::Level::INFO,
	"execute_and_import_block",
	?hash,
	finalized,
	enact_state
);
```

So to just provide aggregates we lose much of the useful information.

Reply:
> am I right in thinking that Histogram does not preserve individual observations?

Yes, that is correct, none of the metric types (counters, gauges, histograms, summary) preserve individual observations and that is very much by design for *metrics* compared to *logging* and *tracing*. *Metrics* make a trade off between the amount of metrics one can collect (within the millions per scrape across ones fleet) and the level of details per observation. I think https://peter.bourgon.org/blog/2017/02/21/metrics-tracing-and-logging.html is a great write-up of the difference for further details.

While this might make the *metrics* effort unsuitable for  a crate called *tracing*, maybe we can introduce an *observability* (the combination of the three - metrics, logging, tracing) in the future, allowing developers to instrument their code easily based on all 3 pillars.

Does all of this make sense @mattrutherford?

Reply:
I agree it's good to be clear on terminology, but that wasn't my main concern - only that if we're ripping out the Grafana server as part of the introduction of this Prometheus feature, is whether we'd lose this functionality to plot individual observations* - however that seems unavoidable based on what you said.

*For my use-case it's OK because I use the Telemetry Receiver in substrate-tracing to send the data to substrate-analytics which uses a PostgreSQL datasource, which in turn can be queried by Grafana (because I want to archive the data); however this is not trivial to set up, so the question is - do we really want to kill substrate Grafana server yet? It's a relatively new feature so maybe not many people use it, but I think it's something we should consider as part of our decision to remove it. Particularly as the current primary use-case for `substrate-tracing` is for profiling and that might be handy for developers to have any easy way to look at this.

Pinging @marcio-diaz @shawntabrizi @DarkEld3r in case you have opinion on this

Reply:
Imo the Grafana server in its current state isn't really _good_ at doing anything. If Prometheus isn't going to work for tracing, we should probably find something that is. I'm happy enough to leave the Grafana server in for the moment until that happens.

Reply:
ok, good to know - and just to be clear that I have no problem removing it, if it's not going to be a big detriment to people - just wanted to make sure everyone was aware of full implications

Reply:
I think tracing other's nodes is not (should not) the purpose of the prometheus exporter. Extracting information from nodes operated by individual entities(validators or full-node operators) is not sounding right as a perspective of information privacy.

I guess tracing functionality should remain in telemetry, and it should be also "off as default" for privacy of node operators.

Reply:
OK - I can see what the misunderstanding is here and it warrants some more explanation - tracing is only capturing local data; which we then have the option to send somewhere, either Log (output tracing data via logger), Grafana server (built-in) or Telemetry. When using telemetry to send the tracing data, it is not expected (or desired) to send to the default telemetry url, so we override that via the cli, with eg: `--telemetry-url 'ws://localhost:8080 9'` to send the data to an analytics server (normally, that we control).